ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java

```diff
@@ -672,6 +676,105 @@ public void ttlUpdateReplicationTest() throws Exception {
+          new MessageInfo(id, putMsgInfo.getFirst().remaining(), 1, accountId, containerId,
+              putMsgInfo.getSecond().getOperationTimeMs()), putMsgInfo.getFirst());
+
+      // add 3 messages to the remote host only
+      expectedIds.addAll(addPutMessagesToReplicasOfPartition(partitionId, Collections.singletonList(remoteHost), 3));
+
+      // delete the very first blob in the remote host only (and delete it from expected list)
+      Iterator<StoreKey> iter = expectedIds.iterator();
+      addDeleteMessagesToReplicasOfPartition(partitionId, iter.next(), Collections.singletonList(remoteHost));
+      iter.remove();
+
+      // PUT and DELETE a blob in the remote host only
+      id = addPutMessagesToReplicasOfPartition(partitionId, Collections.singletonList(remoteHost), 1).get(0);
+      addDeleteMessagesToReplicasOfPartition(partitionId, id, Collections.singletonList(remoteHost));
+
+      idsToExpectByPartition.put(partitionId, expectedIds);
+    }
+
+    // do the exchange metadata.
+
+    StoreKeyFactory storeKeyFactory = new BlobIdFactory(clusterMap);
+    Transformer transformer = new BlobIdTransformer(storeKeyFactory, storeKeyConverter);
+    int batchSize = 400;
+    Pair<Map<DataNodeId, List<RemoteReplicaInfo>>, ReplicaThread> replicasAndThread =
+        getRemoteReplicasAndReplicaThread(batchSize, clusterMap, localHost, remoteHost, storeKeyConverter, transformer,
+            null);
+    Map<DataNodeId, List<RemoteReplicaInfo>> replicasToReplicate = replicasAndThread.getFirst();
+    ReplicaThread replicaThread = replicasAndThread.getSecond();
+
+    // Do the replica metadata exchange.
+    List<ReplicaThread.ExchangeMetadataResponse> responses =
+        replicaThread.exchangeMetadata(new MockConnection(remoteHost, batchSize),
+            replicasToReplicate.get(remoteHost.dataNodeId));
+
+    Assert.assertEquals(
+        idsToExpectByPartition.values().stream().flatMap(Collection::stream).collect(Collectors.toSet()),
+        responses.stream().map(k -> k.missingStoreKeys).flatMap(Collection::stream).collect(Collectors.toSet()));
+
+    // Now delete a message in the remote before doing the Get requests (for every partition). Remove these keys from
+    // expected key set. Even though they are requested, they should not go into the local store. However, this cycle
+    // of replication must be successful
+    for (PartitionId partitionId : partitionIds) {
+      Iterator<StoreKey> iter = idsToExpectByPartition.get(partitionId).iterator();
+      iter.next();
+      StoreKey keyToDelete = iter.next();
+      addDeleteMessagesToReplicasOfPartition(partitionId, keyToDelete, Collections.singletonList(remoteHost));
+      iter.remove();
+    }
+
+    replicaThread.fixMissingStoreKeys(new MockConnection(remoteHost, batchSize),
+        replicasToReplicate.get(remoteHost.dataNodeId), responses);
+
+    Assert.assertEquals(idsToExpectByPartition.keySet(), localHost.infosByPartition.keySet());
```

Comment:
Seemed like a lot of work since `MockTime` cannot be used, but found another way and have added one now. Both tests fail if `Include_None` is used instead of `Include_All`.